sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -128,38 +128,29 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
 
     for (Command *Cmd : AuxiliaryCmds) {
       Enqueued = GraphProcessor::enqueueCommand(Cmd, Res);
-      try {
-        if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
-          throw runtime_error("Auxiliary enqueue process failed.",
-                              PI_INVALID_OPERATION);
-      } catch (...) {
-        // enqueueCommand() func and if statement above may throw an exception,
-        // so destroy required resources to avoid memory leak
-        CleanUp();
-        std::rethrow_exception(std::current_exception());
-      }
+      if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
+        throw runtime_error("Auxiliary enqueue process failed.",
+                            PI_INVALID_OPERATION);
     }
 
     if (NewCmd) {
       // TODO: Check if lazy mode.
       EnqueueResultT Res;
-      try {
-        bool Enqueued = GraphProcessor::enqueueCommand(NewCmd, Res);
-        if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
-          throw runtime_error("Enqueue process failed.", PI_INVALID_OPERATION);
-      } catch (...) {
-        // enqueueCommand() func and if statement above may throw an exception,
-        // so destroy required resources to avoid memory leak
-        CleanUp();
-        std::rethrow_exception(std::current_exception());
-      }
+      bool Enqueued = GraphProcessor::enqueueCommand(NewCmd, Res);
+      if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
+        throw runtime_error("Enqueue process failed.", PI_INVALID_OPERATION);
 
       // If there are no memory dependencies decouple and free the command.
       // Though, dismiss ownership of native kernel command group as it's
       // resources may be in use by backend and synchronization point here is
       // at native kernel execution finish.
       CleanUp();
     }
+  } catch (...) {
+    // If enqueuing has failed we need to clean up the command to remove it
+    // from the graph so it does not cause issues for other related commands.
+    cleanupFinishedCommands(NewEvent);
```

Comment:
Probably it will be slower than CleanUp()

Reply:
It will, but the problem with the current version of `CleanUp()` is that it simply leaves it in the graph if it has either users or dependencies. If we have dependencies, such as accessors and streams, it will have dependencies and stay alive in the graph. This can cause leaks in various implementations, such as `buffer_impl`. The added unittests will fail without these changes.

Consequently `cleanupFinishedCommands` has been adjusted to handle failed commands by simply dropping them. I am not all too familiar with the entire graph structure, so I was hoping you, @romanovvlad and @sergey-semenov could advise on the approach. Can there be a problem with leaving commands user-less? Is there a problem if the failed command has a user and can that even happen here?

Note that the catch here has been moved because deallocation of streams tries to reacquire the lock. So does `cleanupFinishedCommands` but we could work around that.

Reply:
`cleanupFinishedCommands` makes the assumption that all commands in the subgraph have been already finished. This is not guaranteed for this call, and might lead to some issues with some dependencies of the failed command. In particular, I would expect stream deallocation performed in `cleanupFinishedCommands` to trigger a wait for CGs that use them. Although, that might not be as big of a concern if we have a failed enqueue anyway.

Reply:
> `cleanupFinishedCommands` makes the assumption that all commands in the subgraph have been already finished. This is not guaranteed for this call, and might lead to some issues with some dependencies of the failed command.

I see. Would it make sense to add a `cleanupFailedCommand` instead of using an adjusted `cleanupFinishedCommands` for this? It would only clean up the corresponding command and no more. Would it need to do anything special with its users (and would it ever have any?)